### PR TITLE
Add additional telemetry events

### PR DESCRIPTION
**Pull request by dvic**
This PR adds additional telemetry events for 
- `Engine` operations that touch the database
- `Notifier` calls

### CODE REVIEW

File: lib/oban/queue/producer.ex [hunk 1]
```diff
@@ -221,7 +227,13 @@ defmodule Oban.Queue.Producer do
   end
 
   defp start_jobs(%State{} = state) do
-    {:ok, {meta, jobs}} = Engine.fetch_jobs(state.conf, state.meta, state.running)
+    tele_meta = tele_meta(state)
+
+    {meta, jobs} =
```
Comment (sorentwo) [inline lib/oban/queue/producer.ex:232]: This is already covered by the `dispatch` function, though I admit the `[:oban, :producer, _]` namespace isn't very helpful.

**Review by sorentwo — COMMENTED** [review]
There are other db interaction points that don't have any telemetry event around them, e.g. `cancel`, `retry`, `put_meta`. What do you think about wrapping each of the functions in the `Engine` module instead?

File: lib/oban/queue/producer.ex (hunk 1 above)
Comment (dvic) [reply lib/oban/queue/producer.ex:232]: Ah I see, so just `[:oban, :engine, :refresh]` then?

Comment (dvic) [reply lib/oban/queue/producer.ex:232]: Never mind, I misunderstood you, I'll move the Engine events to the engine module.

**Comment by dvic** [comment]
> There are other db interaction points that don't have any telemetry event around them, e.g. `cancel`, `retry`, `put_meta`. What do you think about wrapping each of the functions in the `Engine` module instead?

Yes that seems like a better idea, I'll do that.

By the way: I just pushed also telemetry events for Notifier, need to still update the docs though (will do that once we settle all the events). Any other candidates that you can think of?

**Comment by dvic** [comment]
Currently I'm missing only the following query:

```sql
UPDATE "public"."oban_producers" AS o0 SET "updated_at" = $1 WHERE (o0."uuid" = $2) RETURNING o0."uuid", o0."name", o0."node", o0."queue", o0."meta", o0."running_ids", o0."started_at", o0."updated_at"
```

any idea where this one is coming from?

File: lib/oban/notifier.ex [hunk 2]
```diff
@@ -158,17 +158,29 @@ defmodule Oban.Notifier do
...
     conf = Oban.config(server)
 
-    conf.name
-    |> Registry.whereis(Oban.Notifier)
-    |> conf.notifier.notify(channel, normalize_payload(payload))
+    with_span(conf, channel, payload, fn ->
+      conf.name
+      |> Registry.whereis(Oban.Notifier)
+      |> conf.notifier.notify(channel, normalize_payload(payload))
+    end)
```
Comment (sorentwo) [inline lib/oban/notifier.ex:175]: There was a little duplication in here before, but this amount seems unnecessary. How about this instead?

```elixir
    server
    |> Oban.config()
    |> notify(channel, payload)
```

File: lib/oban/telemetry.ex [hunk 3]
```diff
@@ -69,6 +69,82 @@ defmodule Oban.Telemetry do
   * `:conf` — the config of the Oban supervised producer
   * `:dispatched_count` — the number of jobs fetched and started by the producer
 
+  ### Engine Events
+
+  Oban emits telemetry span events for the following Engine operations:
```
Comment (sorentwo) [inline lib/oban/telemetry.ex:74]: The `init`, `refresh`, and `put_meta` events may also write to the database. So you'll need to wrap those as well to cover all of the interactions.

File: lib/oban/notifier.ex [hunk 4]
```diff
@@ -158,17 +158,29 @@ defmodule Oban.Notifier do
...
+  end
+
+  defp with_span(conf, channel, payload, cb) do
```
Comment (sorentwo) [inline lib/oban/notifier.ex:178]: Very minor, but for the sake of consistency, I'd prefer `fun` rather than `cb` here and in the engine's function.

**Review by sorentwo — CHANGES_REQUESTED** [review]
Nice work. I've left a few nitpicks and pointed out a couple other db interaction spots. After that, this is good to go!

**Comment by dvic** [comment]
> Nice work. I've left a few nitpicks and pointed out a couple other db interaction spots. After that, this is good to go!

Done :)

**Comment by dvic** [comment]
Is there a flaky test? Locally all tests run fine on 1.12 + OTP 24?

**Comment by sorentwo** [comment]
> Is there a flaky test? Locally all tests run fine on 1.12 + OTP 24?

There are some flaky tests due to race conditions and general timing problems. It shows up in CI due to the unpredictable resource contention. I rarely see them locally, and when I do, I fix them.

File: lib/oban/telemetry.ex [hunk 5]
```diff
@@ -69,6 +69,106 @@ defmodule Oban.Telemetry do
...
+  * `[:oban, :engine, :fetch_jobs, :start | :stop | :exception]`
+
+  | event        | measures       | metadata           |
+  | ------------ | -------------- | ------------------ |
+  | `:start`     | `:system_time` | `:conf`, `engine`  |
+  | `:stop`      | `:duration`    | `:conf`, `engine`  |
+  | `:exception` | `:duration`    | `:conf`, `engine`  |
+
+
+  * `[:oban, :engine, :complete_job, :start | :stop | :exception]` — when an engine operation is called
```
Comment (sorentwo) [inline lib/oban/telemetry.ex:109]: This comment would apply to all of the events.

```suggestion
  * `[:oban, :engine, :complete_job, :start | :stop | :exception]`
```

**Review by sorentwo — APPROVED** [review]

**Comment by sorentwo** [comment]
Beautiful work! Thanks 💛

**Pull request merged**